examples/platform/esp32/route_hook/esp_route_hook.c

```diff
@@ -0,0 +1,182 @@
+/*
+ * SPDX-FileCopyrightText: 2015-2021 Espressif Systems (Shanghai) CO LTD
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "esp_route_hook.h"
+
+#include <string.h>
+
+#include "esp_check.h"
+#include "esp_err.h"
+#include "esp_netif.h"
+#include "esp_route_table.h"
+
+#include "lwip/icmp6.h"
+#include "lwip/mld6.h"
+#include "lwip/netif.h"
+#include "lwip/prot/nd6.h"
+#include "lwip/raw.h"
+
+#define HOPLIM_MAX 255
+#define PIO_FLAG_ON_LINK (1 << 7)
+#define PIO_FLAG_AUTO_CONFIG (1 << 6)
+
+#define TAG "ROUTE_HOOK"
+
+typedef struct esp_route_hook
+{
+    struct netif * netif;
+    struct raw_pcb * pcb;
+    struct esp_route_hook * next;
+} esp_route_hook;
+
+PACK_STRUCT_BEGIN
+struct rio_header_t
+{
+    PACK_STRUCT_FLD_8(u8_t type);
+    PACK_STRUCT_FLD_8(u8_t length);
+    PACK_STRUCT_FLD_8(u8_t prefix_length);
+    PACK_STRUCT_FLD_8(u8_t preference);
+    PACK_STRUCT_FIELD(u32_t route_lifetime);
+} PACK_STRUCT_STRUCT;
+PACK_STRUCT_END
+
+typedef struct rio_header_t rio_header_t;
+
+static esp_route_hook * s_hooks;
+
+static bool is_self_address(struct netif * netif, const ip6_addr_t * addr)
+{
+    for (size_t i = 0; i < LWIP_ARRAYSIZE(netif->ip6_addr); i++)
+    {
+        if (ip6_addr_isvalid(netif_ip6_addr_state(netif, i)) &&
+            memcmp(addr->addr, netif_ip6_addr(netif, i)->addr, sizeof(addr->addr)) == 0)
+        {
+            return true;
+        }
+    }
+    return false;
+}
+
+static void ra_recv_handler(struct netif * netif, const uint8_t * icmp_payload, uint16_t payload_len, const ip6_addr_t * src_addr)
+{
+    uint16_t offset = sizeof(struct ra_header);
+
+    while (payload_len >= offset + 2)
+    {
+        uint8_t opt_type = icmp_payload[offset];
+        uint8_t opt_len  = icmp_payload[offset + 1] << 3;
+
+        if (opt_type == ND6_OPTION_TYPE_ROUTE_INFO && !is_self_address(netif, src_addr))
+        {
+            rio_header_t * rio_header = (rio_header_t *) &icmp_payload[offset];
```

Comment:
Before assuming you can cast into payload, you need to check that opt_len would not overrun icmp_payload by checking payload_len. The code here assumes that only size enough for the offset much be checked (2 bytes for next option header) and that the packet is truthful about length of subfields. Please add more size validation before any access to icmp_payload pointer